rasa/core/policies/ensemble.py

```diff
@@ -526,14 +539,22 @@ def _pick_best_policy(
             if form_confidence > best_confidence:
                 best_policy_name = form_policy_name
 
-        return predictions[best_policy_name].probabilities, best_policy_name
+        best_prediction = predictions[best_policy_name]
+
+        # Apply policy events to tracker
+        policy_events += best_prediction.optional_events
+
+        for event in policy_events:
+            tracker.update(event, domain)
```

Comment:
this is still illegal thing to do?

Reply:
Why is it illegal? We add the events **after** all policies made their prediction so they are not affecting each other for the current action.

Reply:
but we update the object that we pass as an input

Reply:
what's the problem with that?

Reply:
you pass one object to the method, but after it you effectively get the different one. I don't know. I thought this was the problem with updating tracker from inside the policies, because there is not so much difference whether it is done in ensemble or policies. cc @tmbo

Reply:
In my opinion it's different if you do it **within** the policies as this might affect the predictions of other policies for the current state. Consider having an ensemble of two policies. The first one is asked to predict the probabilities and adds a featurized `SlotSet` event to the tracker. The second policy would then predict based on a different state. In this PR we update the tracker **after** all predictions were done which will potentially affects the next action prediction but not the current one.

Reply:
with slot set yes, but back then when we had only `FormValidation` that we know doesn't influence any prediction, we had a discussion that tracker shouldn't be updated anywhere outside processor (or whatever class handles adding predicted events), because it is a bad practice from programming point view, because it is preferred to have immutable inputs or whatever is the reason

Reply:
I think it makes sense for the `ensemble` to apply the events because there is some logic to it (e.g. optional events are only applied for the winning policy) which in my opinion is tied to the model (I don't see the `processor` as part of the model). 

> because it is a bad practice from programming point view, because it is preferred to have immutable inputs or whatever is the reason

I don't see the risk in doing so. The `tracker` represents the conversation state so it's normal that that things get appended to it as we go. I can expose the the policy events to the processor but I'm not sure if it makes anything simpler or more transparent.

Reply:
I don't know, I just remember that there was a discussion that we'd like to avoid it. Let's ask in the engineering team what others think.

Reply:
I had another look and it would reduce the amount of individual parameters which we pass around in `processor` by encapsulating everything within a `PolicyPrediction` object which would be desirable. 

@Ghostvv Would we throw away the policy events in case the predicted action was rejected? My intuition says yes.

Reply:
@federicotdn @degiz What do you think? 

Context: `Policy`s can now return `Event`s. Previously only actions and NLU predictions returned events which were applied to the tracker within the `processor` module. I am currently applying the events which are returned by the `Policy`s within the `PolicyEnsemble`. I am indifferent where we apply these events to the tracker. I'm currently slightly leaning towards `Processor` we could use this to reduce the number of parameters in our methods within `processor`. Namely we wouldn't pass around `action_name`, `confidence`, and `policy_name` separately, but could bundle them within one `prediction` argument.

Reply:
yes, if action was rejected, it never happened, therefore all the events which came with its predictions are irrelevant

Reply:
But what if the action needs the policy events? 😆 
This is actually something which requires us to move the application of the events to the tracker to processor.

Reply:
aaah, yeah good catch

Reply:
It was definitely worth doing it. However, the change got bigger than expected. Will work in the remaining todos tomorrow.